File: agents/state.py
"""
In-memory conversation state manager for the Insights agent.
"""

from dataclasses import dataclass, field
from enum import Enum, auto
from typing import Optional


class Phase(Enum):
    IDENTIFY = auto()        # Waiting for user to name a company/role
    CONFIRMING = auto()      # Showing "is this the role you mean?" before synopsis
    AWAITING_SHARE = auto()  # Basic mode: waiting for user to share what they know
    COMPANY_FOUND = auto()   # Synopsis shown, answering follow-ups
    ROLE_FOUND = auto()      # Role synopsis shown, answering follow-ups


@dataclass
class ConversationState:
    user_id: str
    user_name: str = ""
    mode: str = "premium"    # "free", "pro", or "premium"

    phase: Phase = Phase.IDENTIFY

    company_record_id: Optional[str] = None
    company_name: Optional[str] = None
    role_record_id: Optional[str] = None
    role_title: Optional[str] = None

    # Suggested field updates accumulated from the conversation (not written to Airtable)
    suggested_updates: dict = field(default_factory=dict)

    # Full message history for Claude
    messages: list = field(default_factory=list)

    def add_user_message(self, text: str) -> None:
        self.messages.append({"role": "user", "content": text})

    def add_assistant_message(self, text: str) -> None:
        self.messages.append({"role": "assistant", "content": text})


class StateManager:
    def __init__(self):
        self._store: dict[str, ConversationState] = {}

    def get(self, user_id: str) -> Optional[ConversationState]:
        return self._store.get(user_id)

    def get_or_create(self, user_id: str, user_name: str = "", mode: str = "premium") -> ConversationState:
        if user_id not in self._store:
            self._store[user_id] = ConversationState(
                user_id=user_id,
                user_name=user_name,
                mode=mode,
            )
        return self._store[user_id]

    def reset(self, user_id: str, user_name: str = "", mode: str = "premium") -> ConversationState:
        state = ConversationState(user_id=user_id, user_name=user_name, mode=mode)
        self._store[user_id] = state
        return state

    def delete(self, user_id: str) -> None:
        self._store.pop(user_id, None)

File: agents/state.js
'use strict';

/**
 * In-memory conversation state manager for the Insights agent.
 */

const Phase = Object.freeze({
  IDENTIFY: 'IDENTIFY',                        // Waiting for user to name a company/role
  CONFIRMING: 'CONFIRMING',                    // Legacy: kept for safety; no longer entered in normal flow
  DISAMBIGUATING: 'DISAMBIGUATING',            // Multiple role matches — waiting for user to pick one
  DISAMBIGUATING_COMPANY: 'DISAMBIGUATING_COMPANY', // Multiple company matches — waiting for user to pick one
  AWAITING_SHARE: 'AWAITING_SHARE',            // Basic mode: waiting for user to share what they know
  COMPANY_FOUND: 'COMPANY_FOUND',              // Synopsis shown, answering follow-ups
  ROLE_FOUND: 'ROLE_FOUND',                    // Role synopsis shown, answering follow-ups
  COLLECTING_NEW_ENTITY: 'COLLECTING_NEW_ENTITY', // Collecting info to add a new company/role to the DB
});

class ConversationState {
  constructor(userId, userName = '', mode = 'premium') {
    this.userId = userId;
    this.userName = userName;
    this.mode = mode; // 'free', 'pro', or 'premium'

    this.phase = Phase.IDENTIFY;

    this.companyRecordId = null;
    this.companyName = null;
    this.companyDomain = null;  // e.g. 'https://maintainx.com'
    this.roleRecordId = null;
    this.roleTitle = null;
    this.roleAppPage = null;    // e.g. 'https://app.whisperedagent.com/roles/rec...'

    // Role candidates waiting for user disambiguation (list of Airtable record IDs)
    this.candidateRoleIds = [];

    // Company candidates waiting for user disambiguation (list of Airtable record IDs)
    this.candidateCompanyIds = [];

    // Suggested field updates accumulated from the conversation (not written to Airtable)
    this.suggestedUpdates = {};

    // Pending new entity being collected for DB insertion
    // Shape: { companyName, roleTitle, domain, find, notes, step, hasExistingCompany }
    // step: 'confirm' | 'get_company' | 'domain' | 'find' | 'notes'
    this.pendingNewEntity = null;

    // Set to true when user signals "I have a new/another role" so the next role name
    // skips DB lookup and goes straight to the collection flow.
    this.pendingNewRoleSignal = false;

    // Full message history for Claude
    this.messages = [];
  }

  addUserMessage(text) {
    this.messages.push({ role: 'user', content: text });
  }

  addAssistantMessage(text) {
    this.messages.push({ role: 'assistant', content: text });
  }
}

class StateManager {
  constructor() {
    this._store = new Map();
  }

  get(userId) {
    return this._store.get(userId) || null;
  }

  getOrCreate(userId, userName = '', mode = 'premium') {
    if (!this._store.has(userId)) {
      this._store.set(userId, new ConversationState(userId, userName, mode));
    }
    return this._store.get(userId);
  }

  reset(userId, userName = '', mode = 'premium') {
    const state = new ConversationState(userId, userName, mode);
    this._store.set(userId, state);
    return state;
  }

  delete(userId) {
    this._store.delete(userId);
  }
}

module.exports = { Phase, ConversationState, StateManager };
